ssh/agent.go

```diff
@@ -0,0 +1,166 @@
+// Copyright (c) 2020 VMware, Inc. All Rights Reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package ssh
+
+import (
+	"bufio"
+	"fmt"
+	"io"
+	"strings"
+
+	"github.com/pkg/errors"
+	"github.com/sirupsen/logrus"
+	"github.com/vladimirvivien/echo"
+)
+
+// ssh-agent constant identifiers
+const (
+	AgentPidIdentifier = "SSH_AGENT_PID"
+	AuthSockIdentifier = "SSH_AUTH_SOCK"
+)
+
+// singleton instance of agent
+var current *agent
+
+// agentInfo captures the connection information of the ssh-agent
+type agentInfo map[string]string
+
+// agent represents an instance of the ssh-agent
+type agent struct {
+	// pid of the ssh-agent
+	pid string
+
+	// Authentication socket to communicate with the ssh-agent
+	authSockPath string
+
+	// File paths of the keys added to the ssh-agent
+	keyPaths []string
+}
+
+// agentFromInfo parses the information map and returns an instance of agent
+func agentFromInfo(agentInfo agentInfo) *agent {
+	return &agent{
+		pid:          agentInfo[AgentPidIdentifier],
+		authSockPath: agentInfo[AuthSockIdentifier],
+	}
+}
+
+// AddKey adds a key to the ssh-agent process
+func (agent *agent) AddKey(keyPath string) error {
+	e := echo.New()
+	sshAddCmd := e.Prog.Avail("ssh-add")
+	if len(sshAddCmd) == 0 {
+		return errors.New("ssh-add not found")
+	}
+
+	p := e.Env(agent.GetEnvVariables()).
+		RunProc(fmt.Sprintf("ssh-add %s", keyPath))
+	if err := p.Err(); err != nil {
+		return errors.Wrapf(err, "could not add key %s to ssh-agent", keyPath)
+	}
+	agent.keyPaths = append(agent.keyPaths, keyPath)
+	return nil
+}
+
+// Stop removes the keys and kills the ssh-agent process
+func (agent *agent) Stop() error {
+	// set the global variable to nil
+	defer func() {
+		current = nil
```

Comment:
Add a Mutex here to protect that update.

Reply:
Done